compiler/src/main/scala/org/scalajs/core/compiler/JSTreeExtractors.scala

```diff
@@ -14,24 +14,54 @@ import org.scalajs.core.ir.Types._
 object JSTreeExtractors {
 
   object jse {
+
     /**
-     *  A literally named sequence (like in a call to applyDynamicNamed)
+     *  A sequence of tuples (like in a call to applyDynamicNamed)
      *
      *  Example (Scala): method(("name1", x), ("name2", y))
      */
-    object LitNamed {
-      def unapply(exprs: List[Tree]) = unapply0(exprs, Nil)
+    protected trait GenLitNamed {
+
+      protected def failIfNonLit: Boolean
 
       @tailrec
-      private def unapply0(
+      protected final def unapply0(
           exprs: List[Tree],
           acc: List[(StringLiteral, Tree)]
         ): Option[List[(StringLiteral, Tree)]] = exprs match {
         case Tuple2(name: StringLiteral, value) :: xs =>
           unapply0(xs, (name, value) :: acc)
+        case _ :: xs => {
+          if (failIfNonLit)
+            None
+          else
+            unapply0(xs, acc)
+        }
         case Nil => Some(acc.reverse)
-        case _   => None
       }
+
+    }
+
+    /**
+     *  A literally named sequence (like in a call to applyDynamicNamed)
+     *  Where all parameters are expected to be literally named.
+     *
+     *  Example (Scala): method(("name1", x), ("name2", y))
+     */
+    object LitNamed extends GenLitNamed {
```

Comment:
A public object extends a protected trait. That's not nice.

Instead of using inheritance, I would pass `failIfNonLit` as an additional parameter to `unapply0`, and make `GenLitNamed` private.